File: ClaudeRelayAndroid/core-session/src/main/kotlin/relay/session/AuthCoordinator.kt
package relay.session

import kotlinx.coroutines.CompletableDeferred

/**
 * Thrown by a `withAuth` body when the server reports the handler is not
 * authenticated (typically after a transport reconnect leaves the server with
 * a fresh, unauthenticated handler). Maps to Swift's
 * `SessionController.SessionError.isNotAuthenticated` branch
 * (AuthCoordinator.swift:96).
 */
class NotAuthenticatedException : Exception("not authenticated")

/**
 * Owns the authentication surface ported from
 * Sources/ClaudeRelayClient/AuthCoordinator.swift:
 *
 *  - Single-flight: concurrent [ensureAuthenticated] callers share one
 *    in-flight authentication instead of each firing its own `auth_request`
 *    and racing into the server's "Already authenticated" reject path
 *    (AuthCoordinator.swift:36-40, 73-87).
 *  - [withAuth] runs a body with auth ensured and auto-retries the body
 *    exactly once on a [NotAuthenticatedException], re-authenticating first
 *    (AuthCoordinator.swift:92-101).
 *
 * The Swift original threads a `SessionController` through; on Android that
 * coupling lives in the coordinator, so this port is parameterised by injected
 * lambdas — [authenticate] (performs the auth round-trip and updates whatever
 * state [isAuthValid] reads), [isAuthValid] (the local "is the cached auth
 * still valid for the current connection generation" bit), and [resetAuth]
 * (clears the local auth bit so the next [ensureAuthenticated] re-authenticates;
 * parity with Swift's `sessionController?.resetAuth()` at
 * AuthCoordinator.swift:97).
 *
 * Not thread-safe across dispatchers — the Swift original is `@MainActor`
 * isolated. Callers must confine access to a single (main) dispatcher; the
 * single-flight deduplication relies on that confinement.
 */
class AuthCoordinator(
    private val authenticate: suspend () -> Unit,
    private val isAuthValid: () -> Boolean,
    private val resetAuth: () -> Unit = {},
) {
    /**
     * Single-flight handle. Concurrent callers await the same `CompletableDeferred`
     * instead of starting a second authentication (AuthCoordinator.swift:40).
     */
    private var inFlight: CompletableDeferred<Unit>? = null

    /**
     * Ensure the connection is authenticated. Returns immediately when the
     * cached auth is still valid; otherwise performs (or joins) a single
     * in-flight authentication (AuthCoordinator.swift:69-87).
     */
    suspend fun ensureAuthenticated() {
        if (isAuthValid()) return
        inFlight?.let { return it.await() }

        val deferred = CompletableDeferred<Unit>()
        inFlight = deferred
        try {
            authenticate()
            deferred.complete(Unit)
        } catch (t: Throwable) {
            deferred.completeExceptionally(t)
            throw t
        } finally {
            // Mirror Swift's `defer { if authTask == task { authTask = nil } }`:
            // only clear the slot if it still points at this attempt.
            if (inFlight === deferred) inFlight = null
        }
    }

    /**
     * Run [body] with auth ensured. If [body] throws [NotAuthenticatedException]
     * (the server saw a fresh unauthenticated handler after a reconnect),
     * re-authenticate and retry the body exactly once. A second
     * [NotAuthenticatedException] — or any non-auth error — propagates
     * (AuthCoordinator.swift:92-101).
     */
    suspend fun <T> withAuth(body: suspend () -> T): T {
        ensureAuthenticated()
        return try {
            body()
        } catch (e: NotAuthenticatedException) {
            // Swift: `sessionController?.resetAuth(); ensureAuthenticated()`
            // (AuthCoordinator.swift:97-98). resetAuth() clears the local
            // auth bit so the following ensureAuthenticated() can't short-circuit
            // on a stale-valid flag and instead routes through the single-flight
            // `inFlight` slot. Calling ensureAuthenticated() — NOT authenticate()
            // directly — is the load-bearing detail: when two concurrent withAuth
            // bodies both throw NotAuthenticated (the post-reconnect
            // mass-invalidation case), both retries dedupe through the shared
            // in-flight authentication instead of each racing its own
            // auth_request.
            resetAuth()
            ensureAuthenticated()
            body()
        }
    }
}

File: Sources/CodeRelayClient/AuthCoordinator.swift
import Foundation
import CodeRelayKit

/// Owns the authentication surface previously embedded in
/// `SharedSessionCoordinator`:
///
/// - `SessionController` instance bound to the connection
/// - Single-flight `authTask` so concurrent `ensureAuthenticated()` callers
///   share one in-flight `auth_request` instead of racing into "Already
///   authenticated" rejects on the server.
/// - `withAuth { ... }` helper that auto-retries once on stale-auth errors
///   (server replies "Not authenticated" after a transport reconnect).
///
/// The coordinator creates one `AuthCoordinator` per `RelayConnection` + token
/// pair and calls `withAuth { controller in ... }` for every session-lifecycle
/// operation.
@MainActor
public final class AuthCoordinator {

    // MARK: - Dependencies

    public let connection: RelayConnection
    public let token: String

    /// Fires once on each successful authentication. Replaces the previous
    /// `open func didAuthenticate()` subclass hook. Platform subclasses that
    /// used to override `didAuthenticate` now set this closure.
    public var onAuthenticated: (() -> Void)?

    // MARK: - State

    /// The current `SessionController`. Created on first `ensureAuthenticated`
    /// and reused across retries until `resetAuth()` drops it.
    public var sessionController: SessionController?

    /// Single-flight: concurrent callers to `ensureAuthenticated` await the
    /// same in-flight Task. Without this, two coordinated flows (e.g.
    /// scenePhase resume + user tap) each send their own `auth_request` and
    /// the second one hits the server's "Already authenticated" reject path.
    private var authTask: Task<SessionController, Error>?

    // MARK: - Init

    public init(connection: RelayConnection, token: String) {
        self.connection = connection
        self.token = token
    }

    // MARK: - API

    /// True when the cached `sessionController` reports auth valid for the
    /// current `connection.generation`. `forceReconnect()` bumps the
    /// generation; the controller's `isAuthValid` reads through it so the
    /// coordinator can detect the post-reconnect stale-auth state without
    /// having to explicitly invalidate.
    public var isAuthValid: Bool {
        sessionController?.isAuthValid == true
    }

    /// Clear the local auth bit so the next operation re-authenticates. Does
    /// NOT tear down the connection — used when the server replied "Not
    /// authenticated" after a reconnect.
    public func resetAuth() {
        sessionController?.resetAuth()
    }

    /// Return an authenticated `SessionController`. Concurrent callers share
    /// the same in-flight authentication Task.
    public func ensureAuthenticated() async throws -> SessionController {
        if let controller = sessionController, controller.isAuthValid {
            return controller
        }
        if let existing = authTask {
            return try await existing.value
        }
        let task = Task<SessionController, Error> { [weak self] in
            guard let self else { throw SessionController.SessionError.timeout }
            let controller = self.sessionController ?? SessionController(connection: self.connection)
            try await controller.authenticate(token: self.token)
            self.sessionController = controller
            self.onAuthenticated?()
            return controller
        }
        authTask = task
        defer { if authTask == task { authTask = nil } }
        return try await task.value
    }

    /// Run `body` with an authenticated controller. If the body throws an
    /// `isNotAuthenticated`-shaped error (server sees a fresh unauthenticated
    /// handler after a reconnect), reset auth and retry exactly once.
    public func withAuth<T>(_ body: (SessionController) async throws -> T) async throws -> T {
        let controller = try await ensureAuthenticated()
        do {
            return try await body(controller)
        } catch let error as SessionController.SessionError where error.isNotAuthenticated {
            sessionController?.resetAuth()
            let retryController = try await ensureAuthenticated()
            return try await body(retryController)
        }
    }

    /// Cancel any in-flight authentication attempt, but keep the existing
    /// `sessionController` intact. Used by `cancelRecovery` so the user can
    /// reconnect without immediately forcing a fresh `auth_request`.
    public func cancelInFlight() {
        authTask?.cancel()
        authTask = nil
    }

    /// Cancel any in-flight authentication AND drop the cached controller.
    /// Used by `tearDown` when the coordinator is going away permanently.
    public func invalidate() {
        authTask?.cancel()
        authTask = nil
        sessionController = nil
    }
}
